lib/Sema/SemaBounds.cpp

```diff
@@ -3378,6 +3381,335 @@ namespace {
   };
 }
 
+namespace {
+class AvailableFactsAnalysis {
+private:
+  typedef llvm::SmallPtrSet<const Expr *, 16> ExprSet;
+  typedef std::pair<Expr *, Expr *> Comparison;
+  typedef std::set<Comparison> ComparisonSet;
+
+  class ElevatedCFGBlock {
+  private:
+    const CFGBlock *Block;
+    ComparisonSet In, OutThen, OutElse;
+    ComparisonSet Kill, GenThen, GenElse;
+
+  public:
+    ElevatedCFGBlock(const CFGBlock *Block) : Block(Block) {}
+
+    friend class AvailableFactsAnalysis;
```

Comment:
@sunnychatterjee I used `friend` since `AvailableFactsAnalysis` should be the only class that manipulates an object of `ElevatedCFGBlock`. (I thought of it as an Iterator class.)
If I make public methods to access the data members of it, I think I share more than necessary with the world outside `AvailableFactsAnalysis`. Is there a better way to achieve this goal?

Reply:
@ppashakhanloo Thanks for the explanation. I try to avoid "friend" wherever possible since it tends to break encapsulation. However, there are exceptions to this like iterators as you mention. Given that the two classes are so close to each other (nested), let's leave them as is.